### PR TITLE
Optimize access of array member in a structure.

### PR DESCRIPTION
**Pull request by shwqf**
This change adds a simple pattern match for a short AIR sequence of 'struct_field_ptr -> load -> array_elem_val', and generates same LLVM IR as the sequence of 'struct_field_ptr -> ptr_elem_ptr -> load'.

Although this change will still generate CallInst of Memcpy, LLVM will successfully identify the CallInst as dead code and eliminate it, preventing unuseful memcpy of the array field.

Resolves https://github.com/ziglang/zig/issues/13938

### CODE REVIEW

File: test_io_uring_renameat_old [hunk 1]
```diff
@@ -0,0 +1 @@
+hello
```
Comment (matu3ba) [inline test_io_uring_renameat_old:1]: What is this?

**Comment by Vexu** [comment]
It is good to fix things like this but doing it in `llvm.zig` means that it either has to be implemented for every backend or that it needs to be reimplemented in an earlier stage of the compiler.

File: test_io_uring_renameat_old (hunk 1 above)
Comment (shwqf) [reply test_io_uring_renameat_old:1]: .....mistake

**Comment by shwqf** [comment]
> It is good to fix things like this but doing it in `llvm.zig` means that it either has to be implemented for every backend or that it needs to be reimplemented in an earlier stage of the compiler.



> It is good to fix things like this but doing it in `llvm.zig` means that it either has to be implemented for every backend or that it needs to be reimplemented in an earlier stage of the compiler.

I am familiar with LLVM, so I changed llvm.zig first. I will try to optimize it in Sema.zig tomorrow.

**Comment by Vexu** [comment]
> I am familiar with LLVM, so I changed llvm.zig first. I will try to optimize it in Sema.zig tomorrow.

Note that it is not a blocker for this PR since the other backends (besides maybe the C backend) are not ready for use.

**Comment by shwqf** [comment]
> > I am familiar with LLVM, so I changed llvm.zig first. I will try to optimize it in Sema.zig tomorrow.
> 
> Note that it is not a blocker for this PR since the other backends (besides maybe the C backend) are not ready for use.

lol, it would be great if we have a LLVM-like pipeline for AIR

**Comment by IntegratedQuantum** [comment]
> Although this change will still generate CallInst of Memcpy, LLVM will successfully identify the CallInst as dead code and eliminate it

Does that also work in debug mode?
Because if it doesn't and the debug performance is still 10000 times worse, then this PR wouldn't help me(being the author of #13938) since I want to be able to test my program in debug mode.

**Comment by shwqf** [comment]
> > Although this change will still generate CallInst of Memcpy, LLVM will successfully identify the CallInst as dead code and eliminate it
> 
> Does that also work in debug mode? Because if it doesn't and the debug performance is still 10000 times worse, then this PR wouldn't help me(being the author of #13938) since I want to be able to test my program in debug mode.

Now it can

**Comment by matu3ba** [comment]
Can you rebase on top of master and run zig fmt? It fails:
```
-- Install configuration: "Debug"
+ echo Looking for non-conforming code formatting...
Looking for non-conforming code formatting...
+ stage3-debug/bin/zig fmt --check .. --exclude ../test/cases/ --exclude ../build-debug
../src/codegen/llvm.zig
Error: Process completed with exit code 1.
```

**Comment by shwqf** [comment]
> Can you rebase on top of master and run zig fmt? It fails:
> 
> ```
> -- Install configuration: "Debug"
> + echo Looking for non-conforming code formatting...
> Looking for non-conforming code formatting...
> + stage3-debug/bin/zig fmt --check .. --exclude ../test/cases/ --exclude ../build-debug
> ../src/codegen/llvm.zig
> Error: Process completed with exit code 1.
> ```

fixed

**Review by Vexu — APPROVED** [review]

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thank you for sending this enhancement.

In the future, I want to implement #13265. This will make the LLVM backend no longer use the LLVM IR builder API and instead output bitcode directly. Under such circumstances, we will not have access to these C++ APIs, or this SSA structure that LLVM creates. This patch takes us further away from that goal, and so I request that the optimization be done without doing either of these things:

 * Introducing new C++ code
 * Depending on more of LLVM's API surface area

**Comment by shwqf** [comment]
> Thank you for sending this enhancement.
> 
> In the future, I want to implement #13265. This will make the LLVM backend no longer use the LLVM IR builder API and instead output bitcode directly. Under such circumstances, we will not have access to these C++ APIs, or this SSA structure that LLVM creates. This patch takes us further away from that goal, and so I request that the optimization be done without doing either of these things:
> 
> * Introducing new C++ code
> * Depending on more of LLVM's API surface area

The new C++ code is for removing useless memcpy in debug mode, LLVM can optimize it automatically in release mode.  We at least need to introduce 'LLVMInstructionEraseFromParent' for this small optimization.

**Comment by Vexu** [comment]
> * Introducing new C++ code
> 
> * Depending on more of LLVM's API surface area

One possible but slightly harder solution would be to implement this in AstGen.

**Comment by shwqf** [comment]
> > * Introducing new C++ code
> > * Depending on more of LLVM's API surface area
> 
> One possible but slightly harder solution would be to implement this in AstGen.

I tried but I found AST is untyped.

**Comment by Vexu** [comment]
Types are not needed to fix this, field access can result in a pointer via `field_ptr` and array access can take a pointer with `elem_ptr` and then turn it into a value using `load`. The issue is that currently AstGen uses `field_val` and `elem_val` and then relies on LLVM to fix it.

**Comment by shwqf** [comment]
> Types are not needed to fix this, field access can result in a pointer via `field_ptr` and array access can take a pointer with `elem_ptr` and then turn it into a value using `load`. The issue is that currently AstGen uses `field_val` and `elem_val` and then relies on LLVM to fix it.

How can we tell if a field is a **slice** or an **array**?

My first implementation is in AstGen as you suggest, statg3 cannot be compiled.

**Comment by Vexu** [comment]
`elem_ptr` works for all indexable types including arrays and slices.

**Comment by shwqf** [comment]
I got COVID, too tired to continue to finish AstGen version. 

I remove the redundant C++ code, now this PR only work in non-Debug mode.

@Vexu @matu3ba @andrewrk

**Comment by matu3ba** [comment]
Can you rebase on top of master? The tests fail at Autodoc before llvm, so that should fix them.

**Comment by matu3ba** [comment]
I think you need to apply `zig fmt`:
```
+ echo Looking for non-conforming code formatting...
+ stage3-debug/bin/zig fmt --check .. --exclude ../test/cases/ --exclude ../build-debug
Looking for non-conforming code formatting...
../lib/std/c/darwin.zig
Error: Process completed with exit code 1.
```

**Pull request merged**